### PR TITLE
Use Node v4.4.5

### LINKED ISSUE #312: Bundle iojs instead of node v0.10.35?
Currently scripts/download-node.js pulls node v0.10.35 from nodejs.org. This is causing various errors like atom/atom#3783 when it tries to grab a build that doesn't exist (e.g. armv7l) which just returns a 404. (I think this might be causing #282 as well)

I'm guessing it wouldn't _technically_ be too hard to swap out nodejs for iojs (which distributes binaries for a whole pile more architectures) in that script - change a few urls and filenames and it should Just Work for downloading the binary at least. I'd pull req it myself but I don't know what other moving parts there are that it could possibly break and how to test all that.

### LINKED ISSUE #319: Dependencies (package.json): 406 with http/https
I don't know, if this is related to apm. If i'm using http/https for a repo in package.json, the result is always 406.

<strong>is:</strong>
package.json:
```
"dependencies": {
    "tern": "git://github.com/marijnh/tern.git",
  },
```

<strong>should:</strong>
package.json:
```
"dependencies": {
    "tern": "https://github.com/marijnh/tern.git",
  },
```
npm ERR! fetch failed https://github.com/marijnh/tern.git
npm WARN retry will retry, error on last attempt: Error: fetch failed with status code 406
npm ERR! fetch failed https://github.com/marijnh/tern.git
npm WARN retry will retry, error on last attempt: Error: fetch failed with status code 406
npm ERR! fetch failed https://github.com/marijnh/tern.git
npm ERR! Darwin 14.1.0
npm ERR! argv "/Applications/Atom.app/Contents/Resources/app/apm/bin/node" "/Applications/Atom.app/Contents/Resources/app/apm/node_modules/npm/bin/npm-cli.js" "--globalconfig" "/Users/tobiasschubert/.atom/.apm/.apmrc" "--userconfig" "/Users/tobiasschubert/.atom/.apmrc" "install" "--target=0.21.0" "--arch=x64"
npm ERR! node v0.10.35
npm ERR! npm  v2.5.1

npm ERR! fetch failed with status code 406
npm ERR!

### LINKED ISSUE #410: Building with Node 0.12.x makes all apm commands throw errors
Building apm with Node 0.12.x and running any apm command throws this error:

```
dyld: lazy symbol binding failed: Symbol not found: _node_module_register
  Referenced from: /Users/mnq2/code/apm/node_modules/keytar/build/Release/keytar.node
  Expected in: dynamic lookup

dyld: Symbol not found: _node_module_register
  Referenced from: /Users/mnq2/code/apm/node_modules/keytar/build/Release/keytar.node
  Expected in: dynamic lookup

./bin/apm: line 34: 51737 Trace/BPT trap: 5       /Users/mnquintana/code/apm/bin/node --harmony_collections /Users/mnquintana/code/apm/bin/../lib/cli.js list
```

Building with Node 0.10.x however seems to work as expected.

Refs https://github.com/atom/apm/pull/407

### LINKED ISSUE #481: bad option: --harmony_collections
Hi,
Fedora 23 here, Atom 1.3.2, as soon as I run "apm" (just this simple command, no package specified), I get the message:

```bad option: --harmony_collections```

Of course, same error when trying to install a plugin for real.

### LINKED ISSUE #527: Docs improvements
Followed the instructions in the README, because I can't open the Atom app on Mac. 

Running the executable locally doesn't work. 
```
⬡  ./bin/apm config
dyld: lazy symbol binding failed: Symbol not found: _node_module_register
  Referenced from: /Users/kevinSuttle/Code/apm-master/node_modules/git-utils/build/Release/git.node
  Expected in: dynamic lookup

dyld: Symbol not found: _node_module_register
  Referenced from: /Users/kevinSuttle/Code/apm-master/node_modules/git-utils/build/Release/git.node
  Expected in: dynamic lookup

./bin/apm: line 28: 15735 Trace/BPT trap: 5       "$binDir/$nodeBin" --harmony_collections "$binDir/../lib/cli.js" "$@"
```

I tried making an alias. That didn't work. 

Tried making a symbolic link. That didn't work either.

Tried moving it. That didn't work either.
```
⬡  apm
/usr/local/bin/node: bad option: --harmony_collections
```

Tried re-building it. Nope. `apm` is here in the repo. Not sure what we're rebuilding, then.

### PR DESCRIPTION
**Pull request by BinaryMuse**
This PR updates the bundled version of Node that ships with apm to v4.4.5, the current LTS build. It also changes up the build process considerably to facilitate running with this version of Node on GitHub's CI.

* The bundled version of Node is specified by the contents of the `BUNDLED_NODE_VERSION` file.
* There is a `build/` directory, similar to the one atom/atom uses, that includes dependencies that are necessary for bootstrapping the build — namely, for running the `download-node.js` script which grabs the executable for the bundled Node and puts it in `bin/`.
* There is a `script/install.sh`/`script/install.cmd` script that should be used to install apm when developing on it (instead of doing a plain `npm install`).
* There is a `bin/apm[.cmd]` script that should be used to ensure npm scripts run with the proper version of Node. This can be skipped if you're developing apm using the same version of Node as is listed in `BUNDLED_NODE_VERSION`.

It's likely that we'll need to make some small tweaks to Atom's build steps to make this all work, and I'd like to run it on my own machine for a bit, but I wanted to get this out as early as possible to get some 👀  on it. One thing I'm particularly interested in people's opinion on is if we should stick to a LTS version of Node or if we should jump straight into Node v6.

Fixes #410
Fixes #481
Fixes #527
Fixes #457
Fixes #312
Fixes #319

### CODE REVIEW

**Comment by iolsen** [comment]
Considering how long it took us to take on this upgrade, I'm 👍 on choosing the LTS version.

**Comment by BinaryMuse** [comment]
@iolsen Ha, well the problem was with CI. If this successfully takes care of that problem, we should be able to keep this up-to-date much more easily.

**Comment by 50Wliu** [comment]
Just going to link a "few" issues that should be resolved by this upgrade:
#410
#481
#527
#457
#322
#312
#319 (maybe?  This looks like it has already been resolved)
#328 kinda
#463
#436

File: script/download-node.js [hunk 1]
```diff
@@ -1,75 +1,58 @@
...
   });
 };
 
 var copyNodeBinToLocation = function(callback, version, targetFilename, fromDirectory) {
   var arch = process.arch === 'ia32' ? 'x86' : process.arch;
   var subDir = "node-" + version + "-" + process.platform + "-" + arch;
-  var fromPath = path.join(fromDirectory, subDir, 'bin', 'node');
-  return mv(fromPath, targetFilename, function(err) {
+  var downloadedNodePath = path.join(fromDirectory, subDir, 'bin', 'node');
+  var targetDir = path.dirname(targetFilename)
```
Comment (as-cii) [inline script/download-node.js]: Super minor, but can we 🔥 this line? Seems like we're not using it anywhere in this file.

Comment (BinaryMuse) [inline script/download-node.js]: Good catch! Missed some stuff moving junk around :)

**Comment by BinaryMuse** [comment]
@kuychaco This is now ready as well. Care to give it one more 👀 ?

**Comment by BinaryMuse** [comment]
I think we should merge #562 first, and then I'll rebase this on top of it.

**Comment by NicoHood** [comment]
Does this solve anything of this? https://github.com/atom/atom/issues/7822

**Comment by BinaryMuse** [comment]
@NicoHood My understanding is that it could, but I don't know for usre.

**Comment by iolsen** [comment]
This should also fix (or make possible a fix for) atom/atom#5109! 🎉

**Comment by tiangolo** [comment]
When will this be integrated in Atom? Or where should I check?

I just installed the latest stable version V1.8.0, but I don't even know if this is included in a beta or where to check.

I'm having issues installing a package that uses (`zmq-prebuilt`) and it requires a Node version higher than 0.12.

**Comment by iolsen** [comment]
Soon. We're working on updates to build infrastructure to finish this.

You can check the [Atom releases page](https://github.com/atom/atom/releases) for PRs that went into each version.

**Comment by iolsen** [comment]
Soon. We're working on build infrastructure to include this.

You can look at the [Atom releases page](https://github.com/atom/atom/releases) to see PRs included in each release. You can also do `apm --version` to show the versions of node and npm in use.

**Comment by tiangolo** [comment]
OK, thanks @iolsen.

**Comment by tinymins** [comment]
image
Still get error `node.exe: bad option: --harmony_collections`.
What should I do to fix this problem?

**Comment by ntomallen** [comment]
I'm getting [issues ](https://github.com/ricardofbarros/linter-js-standard/issues/115) installing a linter-js-standard for Atom. It's been suggested that this update (with more recent node/npm versions) will fix the issue. Just curious when this will be updated in the release for Windows. Is it in the Windows beta channel yet?

**Comment by as-cii** [comment]
@ntomallen: this hasn't been released yet, but it will go out on the next beta release! 🚀

**Comment by ntomallen** [comment]
Fantastic! Will definitely be testing it then.

**Pull request merged**